pkg/sql/importer/read_import_csv.go

```diff
@@ -221,11 +221,18 @@ func (c *csvRowConsumer) FillDatums(
 			var err error
 			conv.Datums[datumIdx], err = rowenc.ParseDatumStringAs(conv.VisibleColTypes[i], field.Val, conv.EvalCtx)
 			if err != nil {
-				col := conv.VisibleCols[i]
-				return newImportRowError(
-					errors.Wrapf(err, "parse %q as %s", col.GetName(), col.GetType().SQLString()),
-					strRecord(record, c.opts.Comma),
-					rowNum)
+				// Fallback to parsing as a string literal. This allows us to support
+				// both array expressions (like `ARRAY[1, 2, 3]`) and literals (like
+				// `{1, 2, 3}`).
+				var err2 error
+				conv.Datums[datumIdx], _, err2 = tree.ParseAndRequireString(conv.VisibleColTypes[i], field.Val, conv.EvalCtx)
```

Comment:
It is probably outside the scope of this change, but is there a grammar that a parser can be generated from?

Reply:
you mean to avoid the fallback logic? it won't be that easy -- the difference between these two formats is that one is an expression (e.g. `ARRAY[1, 2, 3 ...]`) but the other is the format that arrays represented as string literals use (e.g. in `{1, 2, 3 ... }::INT[]`). The string literal format isn't (and i think shouldn't) be represented in our sql.y file, but it's handled in pkg/sql/sem/tree/parse_array.go